File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.LinkedList;
import java.util.Queue;

class Point{
    public int x,y;
    Point(int x, int y){
        this.x = x;
        this.y = y;
    }
}

class Solution {
    public int BFS(int[][] maps, int x, int y) {
        int xLength = maps.length;
        int yLength = maps[0].length;
        boolean visited = false;
        
        Queue<Point> q = new LinkedList<>(); 
        q.offer(new Point(x,y));
        while(!q.isEmpty()){
            Point cur = q.poll();
            if(cur.x == xLength - 1 && cur.y == yLength-1){
                visited = true;
            }
            for(int i=0; i<4; i++){
                int nx = cur.x + dx[i];
                int ny = cur.y + dy[i];
                if(0<= nx && nx < xLength && 0<= ny && ny < yLength && maps[nx][ny] == 1){
                    maps[nx][ny] = maps[cur.x][cur.y] + 1;
                    q.offer(new Point(nx,ny));
                }
            }
        }
        
        if(visited){
            return maps[maps.length-1][maps[0].length-1];
        }
        else{
            return -1;
        }
        
    }
    static int[] dx = {0, 0, 1, -1};
    static int[] dy = {1, -1, 0, 0};
    
    
    public int solution(int[][] maps) {
        int answer = BFS(maps, 0, 0);
        return answer;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var answer = 0;
    var [n,m] = [maps.length, maps[0].length];
    
    var dx = [1,0,0,-1];
    var dy = [0,1,-1,0];
    
    var distance = [];
    for(let i=0;i<n;i++){
        distance.push(new Array(m).fill(Infinity))
    }
    
    distance[0][0] = 1;
    
    var queue = [[0,0]];
    
    while(queue.length !== 0){
        
        if(distance[n-1][m-1] !== Infinity){
            return distance[n-1][m-1];
        }
        
        [cx,cy] = queue.shift()
        
        for(let i=0;i<4;i++){
            [nx,ny] = [Number(cx + dx[i]),Number(cy + dy[i])];
            
            if(0<= nx && nx<n && 0<=ny && ny<m && maps[nx][ny] === 1 && distance[cx][cy] + 1 < distance[nx][ny]){
                
                
                distance[nx][ny] = distance[cx][cy] + 1;
                queue.push([nx,ny]);
            }  
        }
    }
    
    return -1;
    
}